File: sdks/typescript/packages/core/src/http/streamingClient.ts
import { T402PaymentError } from "../errors";

const DEFAULT_FACILITATOR_URL = "https://facilitator.t402.io";

// ============================================================================
// Streaming Types
// ============================================================================

export type StreamStatus =
  | "pending"
  | "active"
  | "paused"
  | "closing"
  | "closed"
  | "cancelled"
  | "expired";

export interface StreamMetadata {
  resourceId?: string;
  description?: string;
  tags?: string[];
  extra?: Record<string, string>;
}

export interface Stream {
  id: string;
  network: string;
  scheme: string;
  payer: string;
  payee: string;
  asset: string;
  maxAmount: string;
  currentAmount: string;
  settledAmount: string;
  ratePerSecond: string;
  status: StreamStatus;
  createdAt: string;
  activatedAt?: string;
  lastUpdatedAt: string;
  expiresAt?: string;
  closedAt?: string;
  depositTxHash: string;
  settlementTxHash: string;
  metadata: StreamMetadata;
}

export interface StreamUpdate {
  id: string;
  streamId: string;
  amount: string;
  signature: string;
  timestamp: string;
  sequenceNum: number;
  resourceUnits: number;
}

export interface StreamStats {
  totalUpdates: number;
  averageRate: string;
  duration: number;
  resourcesUsed: number;
  efficiencyScore: number;
}

// Request types

export interface OpenStreamRequest {
  network: string;
  scheme: string;
  payer: string;
  payee: string;
  asset: string;
  maxAmount: string;
  ratePerSecond?: string;
  expiresAt?: string;
  signature: string;
  metadata?: StreamMetadata;
}

export interface UpdateStreamRequest {
  streamId: string;
  amount: string;
  signature: string;
  resourceUnits?: number;
}

export interface CloseStreamRequest {
  streamId: string;
  finalAmount: string;
  payerSignature: string;
  payeeSignature?: string;
  reason?: string;
}

export interface ListStreamsParams {
  network?: string;
  payer?: string;
  payee?: string;
  status?: StreamStatus[];
  limit?: number;
  offset?: number;
  orderBy?: string;
  orderDesc?: boolean;
}

// Response types

export interface OpenStreamResponse {
  stream: Stream;
  depositAmount?: string;
  depositTo?: string;
}

export interface UpdateStreamResponse {
  stream: Stream;
  update: StreamUpdate;
  remaining: string;
  canContinue: boolean;
}

export interface CloseStreamResponse {
  stream: Stream;
  settledAmount: string;
  txHash: string;
  refundAmount: string;
}

export interface GetStreamResponse {
  stream: Stream;
  updates?: StreamUpdate[];
  stats?: StreamStats;
}

export interface ListStreamsResponse {
  streams: Stream[];
  total: number;
  limit: number;
  offset: number;
  hasMore: boolean;
}

export interface PauseResumeResponse {
  status: string;
  message: string;
}

// ============================================================================
// Streaming Client Config
// ============================================================================

export interface StreamingClientConfig {
  url?: string;
  apiKey?: string;
  requesterAddress?: string;
}

// ============================================================================
// Streaming Client
// ============================================================================

/**
 * HTTP client for the t402 facilitator streaming API.
 * Provides methods for managing payment streams.
 */
export class StreamingClient {
  readonly url: string;
  private readonly apiKey?: string;
  private readonly requesterAddress?: string;

  /**
   * Create a new StreamingClient instance.
   *
   * @param config - Client configuration options
   */
  constructor(config?: StreamingClientConfig) {
    this.url = config?.url || DEFAULT_FACILITATOR_URL;
    this.apiKey = config?.apiKey;
    this.requesterAddress = config?.requesterAddress;
  }

  /**
   * Open a new payment stream.
   *
   * @param params - Stream opening parameters
   * @returns The opened stream and deposit details
   */
  async openStream(params: OpenStreamRequest): Promise<OpenStreamResponse> {
    return this.post<OpenStreamResponse>("/v1/stream/open", params);
  }

  /**
   * Update a stream with a new cumulative amount.
   *
   * @param params - Stream update parameters
   * @returns The updated stream state and remaining balance
   */
  async updateStream(params: UpdateStreamRequest): Promise<UpdateStreamResponse> {
    return this.post<UpdateStreamResponse>("/v1/stream/update", params);
  }

  /**
   * Close a stream and settle the final amount.
   *
   * @param params - Stream closing parameters
   * @returns The closed stream with settlement details
   */
  async closeStream(params: CloseStreamRequest): Promise<CloseStreamResponse> {
    return this.post<CloseStreamResponse>("/v1/stream/close", params);
  }

  /**
   * Get stream details by ID.
   *
   * @param id - The stream identifier
   * @param options - Optional query parameters
   * @param options.includeUpdates - Whether to include stream updates
   * @param options.includeStats - Whether to include stream statistics
   * @returns The stream details with optional updates and stats
   */
  async getStream(
    id: string,
    options?: { includeUpdates?: boolean; includeStats?: boolean },
  ): Promise<GetStreamResponse> {
    const params = new URLSearchParams();
    if (options?.includeUpdates) params.set("include_updates", "true");
    if (options?.includeStats) params.set("include_stats", "true");
    const query = params.toString();
    const path = `/v1/stream/${encodeURIComponent(id)}${query ? `?${query}` : ""}`;
    return this.get<GetStreamResponse>(path);
  }

  /**
   * Pause an active stream.
   *
   * @param id - The stream identifier
   * @returns The pause status and message
   */
  async pauseStream(id: string): Promise<PauseResumeResponse> {
    return this.post<PauseResumeResponse>(`/v1/stream/${encodeURIComponent(id)}/pause`, {});
  }

  /**
   * Resume a paused stream.
   *
   * @param id - The stream identifier
   * @returns The resume status and message
   */
  async resumeStream(id: string): Promise<PauseResumeResponse> {
    return this.post<PauseResumeResponse>(`/v1/stream/${encodeURIComponent(id)}/resume`, {});
  }

  /**
   * List streams with optional filters.
   *
   * @param filters - Optional filtering and pagination parameters
   * @returns A paginated list of streams
   */
  async listStreams(filters?: ListStreamsParams): Promise<ListStreamsResponse> {
    const params = new URLSearchParams();
    if (filters?.network) params.set("network", filters.network);
    if (filters?.payer) params.set("payer", filters.payer);
    if (filters?.payee) params.set("payee", filters.payee);
    if (filters?.status) {
      for (const s of filters.status) params.append("status", s);
    }
    if (filters?.limit !== undefined) params.set("limit", String(filters.limit));
    if (filters?.offset !== undefined) params.set("offset", String(filters.offset));
    if (filters?.orderBy) params.set("orderBy", filters.orderBy);
    if (filters?.orderDesc) params.set("orderDesc", "true");
    const query = params.toString();
    return this.get<ListStreamsResponse>(`/v1/stream${query ? `?${query}` : ""}`);
  }

  // ============================================================================
  // Internal HTTP helpers
  // ============================================================================

  /**
   * Build HTTP headers for API requests.
   *
   * @returns A headers object with auth and content-type
   */
  private buildHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };
    if (this.apiKey) {
      headers["Authorization"] = `Bearer ${this.apiKey}`;
    }
    if (this.requesterAddress) {
      headers["X-Requester-Address"] = this.requesterAddress;
    }
    return headers;
  }

  /**
   * Send a POST request to the streaming API.
   *
   * @param path - The API endpoint path
   * @param body - The request body to send as JSON
   * @returns The parsed JSON response
   */
  private async post<T>(path: string, body: unknown): Promise<T> {
    const response = await fetch(`${this.url}${path}`, {
      method: "POST",
      headers: this.buildHeaders(),
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const errorText = await response.text().catch(() => response.statusText);
      throw new T402PaymentError(
        `Streaming API ${path} failed (${response.status}): ${errorText}`,
        {
          phase: "unknown",
          code: response.status,
          retryable: response.status >= 500 || response.status === 429,
        },
      );
    }

    return (await response.json()) as T;
  }

  /**
   * Send a GET request to the streaming API.
   *
   * @param path - The API endpoint path
   * @returns The parsed JSON response
   */
  private async get<T>(path: string): Promise<T> {
    const response = await fetch(`${this.url}${path}`, {
      method: "GET",
      headers: this.buildHeaders(),
    });

    if (!response.ok) {
      const errorText = await response.text().catch(() => response.statusText);
      throw new T402PaymentError(
        `Streaming API ${path} failed (${response.status}): ${errorText}`,
        {
          phase: "unknown",
          code: response.status,
          retryable: response.status >= 500 || response.status === 429,
        },
      );
    }

    return (await response.json()) as T;
  }
}

File: sdks/go/http/streaming_client.go
package http

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

// ============================================================================
// Streaming Types
// ============================================================================

// StreamStatus represents the current status of a payment stream
type StreamStatus string

const (
	StreamStatusPending   StreamStatus = "pending"
	StreamStatusActive    StreamStatus = "active"
	StreamStatusPaused    StreamStatus = "paused"
	StreamStatusClosing   StreamStatus = "closing"
	StreamStatusClosed    StreamStatus = "closed"
	StreamStatusCancelled StreamStatus = "cancelled"
	StreamStatusExpired   StreamStatus = "expired"
)

// StreamMetadata holds optional stream metadata
type StreamMetadata struct {
	ResourceID  string            `json:"resourceId,omitempty"`
	Description string            `json:"description,omitempty"`
	Tags        []string          `json:"tags,omitempty"`
	Extra       map[string]string `json:"extra,omitempty"`
}

// Stream represents a payment stream/channel
type Stream struct {
	ID               string         `json:"id"`
	Network          string         `json:"network"`
	Scheme           string         `json:"scheme"`
	Payer            string         `json:"payer"`
	Payee            string         `json:"payee"`
	Asset            string         `json:"asset"`
	MaxAmount        string         `json:"maxAmount"`
	CurrentAmount    string         `json:"currentAmount"`
	SettledAmount    string         `json:"settledAmount"`
	RatePerSecond    string         `json:"ratePerSecond"`
	Status           StreamStatus   `json:"status"`
	CreatedAt        time.Time      `json:"createdAt"`
	ActivatedAt      *time.Time     `json:"activatedAt"`
	LastUpdatedAt    time.Time      `json:"lastUpdatedAt"`
	ExpiresAt        *time.Time     `json:"expiresAt"`
	ClosedAt         *time.Time     `json:"closedAt"`
	DepositTxHash    string         `json:"depositTxHash"`
	SettlementTxHash string         `json:"settlementTxHash"`
	Metadata         StreamMetadata `json:"metadata"`
}

// StreamUpdate represents an update to a stream
type StreamUpdate struct {
	ID            string    `json:"id"`
	StreamID      string    `json:"streamId"`
	Amount        string    `json:"amount"`
	Signature     string    `json:"signature"`
	Timestamp     time.Time `json:"timestamp"`
	SequenceNum   uint64    `json:"sequenceNum"`
	ResourceUnits int64     `json:"resourceUnits"`
}

// StreamStats contains statistics about a stream
type StreamStats struct {
	TotalUpdates    int64   `json:"totalUpdates"`
	AverageRate     string  `json:"averageRate"`
	Duration        int64   `json:"duration"`
	ResourcesUsed   int64   `json:"resourcesUsed"`
	EfficiencyScore float64 `json:"efficiencyScore"`
}

// OpenStreamRequest is the request to open a new stream
type OpenStreamRequest struct {
	Network       string          `json:"network"`
	Scheme        string          `json:"scheme"`
	Payer         string          `json:"payer"`
	Payee         string          `json:"payee"`
	Asset         string          `json:"asset"`
	MaxAmount     string          `json:"maxAmount"`
	RatePerSecond string          `json:"ratePerSecond,omitempty"`
	ExpiresAt     *time.Time      `json:"expiresAt,omitempty"`
	Signature     string          `json:"signature"`
	Metadata      *StreamMetadata `json:"metadata,omitempty"`
}

// OpenStreamResponse is the response after opening a stream
type OpenStreamResponse struct {
	Stream        *Stream `json:"stream"`
	DepositAmount string  `json:"depositAmount,omitempty"`
	DepositTo     string  `json:"depositTo,omitempty"`
}

// UpdateStreamRequest is the request to update a stream
type UpdateStreamRequest struct {
	StreamID      string `json:"streamId"`
	Amount        string `json:"amount"`
	Signature     string `json:"signature"`
	ResourceUnits int64  `json:"resourceUnits,omitempty"`
}

// UpdateStreamResponse is the response after updating a stream
type UpdateStreamResponse struct {
	Stream      *Stream       `json:"stream"`
	Update      *StreamUpdate `json:"update"`
	Remaining   string        `json:"remaining"`
	CanContinue bool          `json:"canContinue"`
}

// CloseStreamRequest is the request to close a stream
type CloseStreamRequest struct {
	StreamID       string `json:"streamId"`
	FinalAmount    string `json:"finalAmount"`
	PayerSignature string `json:"payerSignature"`
	PayeeSignature string `json:"payeeSignature,omitempty"`
	Reason         string `json:"reason,omitempty"`
}

// CloseStreamResponse is the response after closing a stream
type CloseStreamResponse struct {
	Stream        *Stream `json:"stream"`
	SettledAmount string  `json:"settledAmount"`
	TxHash        string  `json:"txHash"`
	RefundAmount  string  `json:"refundAmount"`
}

// GetStreamOptions configures optional query parameters for GetStream
type GetStreamOptions struct {
	IncludeUpdates bool
	IncludeStats   bool
}

// GetStreamResponse is the response for getting stream details
type GetStreamResponse struct {
	Stream  *Stream         `json:"stream"`
	Updates []*StreamUpdate `json:"updates,omitempty"`
	Stats   *StreamStats    `json:"stats,omitempty"`
}

// PauseResumeResponse is the response for pause/resume operations
type PauseResumeResponse struct {
	Status  string `json:"status"`
	Message string `json:"message"`
}

// ListStreamsParams configures optional filters for listing streams
type ListStreamsParams struct {
	Network   string
	Payer     string
	Payee     string
	Status    []StreamStatus
	Limit     int
	Offset    int
	OrderBy   string
	OrderDesc bool
}

// ListStreamsResponse is the response for listing streams
type ListStreamsResponse struct {
	Streams []*Stream `json:"streams"`
	Total   int64     `json:"total"`
	Limit   int       `json:"limit"`
	Offset  int       `json:"offset"`
	HasMore bool      `json:"hasMore"`
}

// ============================================================================
// Streaming Client
// ============================================================================

// StreamingClientConfig configures the streaming client
type StreamingClientConfig struct {
	// URL is the base URL of the facilitator service
	URL string

	// HTTPClient is the HTTP client to use (optional)
	HTTPClient *http.Client

	// APIKey for authentication (optional)
	APIKey string

	// RequesterAddress for pause/resume operations (optional)
	RequesterAddress string

	// Timeout for requests (optional, defaults to 30s)
	Timeout time.Duration
}

// StreamingClient communicates with the facilitator streaming API
type StreamingClient struct {
	url              string
	httpClient       *http.Client
	apiKey           string
	requesterAddress string
}

// NewStreamingClient creates a new streaming client
func NewStreamingClient(config *StreamingClientConfig) *StreamingClient {
	if config == nil {
		config = &StreamingClientConfig{}
	}

	u := config.URL
	if u == "" {
		u = DefaultFacilitatorURL
	}

	httpClient := config.HTTPClient
	if httpClient == nil {
		timeout := config.Timeout
		if timeout == 0 {
			timeout = 30 * time.Second
		}
		httpClient = &http.Client{Timeout: timeout}
	}

	return &StreamingClient{
		url:              u,
		httpClient:       httpClient,
		apiKey:           config.APIKey,
		requesterAddress: config.RequesterAddress,
	}
}

// OpenStream opens a new payment stream
func (c *StreamingClient) OpenStream(ctx context.Context, req *OpenStreamRequest) (*OpenStreamResponse, error) {
	var resp OpenStreamResponse
	if err := c.doPost(ctx, "/v1/stream/open", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// UpdateStream updates a stream with a new cumulative amount
func (c *StreamingClient) UpdateStream(ctx context.Context, req *UpdateStreamRequest) (*UpdateStreamResponse, error) {
	var resp UpdateStreamResponse
	if err := c.doPost(ctx, "/v1/stream/update", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// CloseStream closes a stream and settles the final amount
func (c *StreamingClient) CloseStream(ctx context.Context, req *CloseStreamRequest) (*CloseStreamResponse, error) {
	var resp CloseStreamResponse
	if err := c.doPost(ctx, "/v1/stream/close", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// GetStream gets stream details by ID
func (c *StreamingClient) GetStream(ctx context.Context, id string, opts *GetStreamOptions) (*GetStreamResponse, error) {
	params := url.Values{}
	if opts != nil {
		if opts.IncludeUpdates {
			params.Set("include_updates", "true")
		}
		if opts.IncludeStats {
			params.Set("include_stats", "true")
		}
	}
	path := "/v1/stream/" + url.PathEscape(id)
	if query := params.Encode(); query != "" {
		path += "?" + query
	}

	var resp GetStreamResponse
	if err := c.doGet(ctx, path, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// PauseStream pauses an active stream
func (c *StreamingClient) PauseStream(ctx context.Context, id string) (*PauseResumeResponse, error) {
	var resp PauseResumeResponse
	path := "/v1/stream/" + url.PathEscape(id) + "/pause"
	if err := c.doPost(ctx, path, struct{}{}, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// ResumeStream resumes a paused stream
func (c *StreamingClient) ResumeStream(ctx context.Context, id string) (*PauseResumeResponse, error) {
	var resp PauseResumeResponse
	path := "/v1/stream/" + url.PathEscape(id) + "/resume"
	if err := c.doPost(ctx, path, struct{}{}, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// ListStreams lists streams with optional filters
func (c *StreamingClient) ListStreams(ctx context.Context, filters *ListStreamsParams) (*ListStreamsResponse, error) {
	params := url.Values{}
	if filters != nil {
		if filters.Network != "" {
			params.Set("network", filters.Network)
		}
		if filters.Payer != "" {
			params.Set("payer", filters.Payer)
		}
		if filters.Payee != "" {
			params.Set("payee", filters.Payee)
		}
		for _, s := range filters.Status {
			params.Add("status", string(s))
		}
		if filters.Limit > 0 {
			params.Set("limit", strconv.Itoa(filters.Limit))
		}
		if filters.Offset > 0 {
			params.Set("offset", strconv.Itoa(filters.Offset))
		}
		if filters.OrderBy != "" {
			params.Set("orderBy", filters.OrderBy)
		}
		if filters.OrderDesc {
			params.Set("orderDesc", "true")
		}
	}
	path := "/v1/stream"
	if query := params.Encode(); query != "" {
		path += "?" + query
	}

	var resp ListStreamsResponse
	if err := c.doGet(ctx, path, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// ============================================================================
// Internal HTTP helpers
// ============================================================================

func (c *StreamingClient) setHeaders(req *http.Request) {
	req.Header.Set("Content-Type", "application/json")
	if c.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.apiKey)
	}
	if c.requesterAddress != "" {
		req.Header.Set("X-Requester-Address", c.requesterAddress)
	}
}

func (c *StreamingClient) doPost(ctx context.Context, path string, body interface{}, result interface{}) error {
	bodyBytes, err := json.Marshal(body)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", c.url+path, bytes.NewReader(bodyBytes))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}
	c.setHeaders(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("streaming API %s request failed: %w", path, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		respBody, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("streaming API %s failed (%d): %s", path, resp.StatusCode, string(respBody))
	}

	if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
		return fmt.Errorf("failed to decode response: %w", err)
	}
	return nil
}

func (c *StreamingClient) doGet(ctx context.Context, path string, result interface{}) error {
	req, err := http.NewRequestWithContext(ctx, "GET", c.url+path, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}
	c.setHeaders(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("streaming API %s request failed: %w", path, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		respBody, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("streaming API %s failed (%d): %s", path, resp.StatusCode, string(respBody))
	}

	if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
		return fmt.Errorf("failed to decode response: %w", err)
	}
	return nil
}
